src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs

```diff
@@ -0,0 +1,214 @@
+    ///
+    /// ### Example
+    /// ```rust
+    /// use std::sync::Mutex;
+    ///
+    /// struct State {}
+    ///
+    /// impl State {
+    ///     fn foo(&self) -> bool {
+    ///         true
+    ///     }
+    ///
+    ///     fn bar(&self) {}
+    /// }
+    ///
+    /// fn main() {
+    ///     let mutex = Mutex::new(State {});
+    ///
+    ///     match mutex.lock().unwrap().foo() {
+    ///         true => {
+    ///             mutex.lock().unwrap().bar(); // Deadlock!
+    ///         }
+    ///         false => {}
+    ///     };
+    ///
+    ///     println!("All done!");
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// // example code which does not raise clippy warning
+    /// use std::sync::Mutex;
+    ///
+    /// struct State {}
+    ///
+    /// impl State {
+    ///     fn foo(&self) -> bool {
+    ///         true
+    ///     }
+    ///
+    ///     fn bar(&self) {}
+    /// }
+    ///
+    /// fn main() {
+    ///     let mutex = Mutex::new(State {});
+    ///
+    ///     let is_foo = mutex.lock().unwrap().foo();
+    ///     match is_foo {
+    ///         true => {
+    ///             mutex.lock().unwrap().bar();
+    ///         }
+    ///         false => {}
+    ///     };
+    ///
+    ///     println!("All done!");
+    /// }
+    /// ```
+    #[clippy::version = "1.60.0"]
+    pub SIGNIFICANT_DROP_IN_SCRUTINEE,
+    nursery,
+    "warns when a temporary of a type with a drop with a significant side-effect might have a surprising lifetime"
+}
+
+declare_lint_pass!(SignificantDropInScrutinee => [SIGNIFICANT_DROP_IN_SCRUTINEE]);
+
+impl<'tcx> LateLintPass<'tcx> for SignificantDropInScrutinee {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        if let Some(expr_span) = has_significant_drop_in_scrutinee(cx, expr) {
+            span_lint_and_help(
+                cx,
+                SIGNIFICANT_DROP_IN_SCRUTINEE,
+                expr_span,
+                "The lifetimes of temporaries in a match scrutinee last to the end of the \
+                match expressions. One or more of the temporaries in this match scrutinee have a \
+                drop with a side-effect (such as unlocking a mutex). Consider calling the scrutinee \
+                outside of the match expression and using the result instead.",
+                None,
+                format!(
+                    "Try this:\nlet value = {};\nmatch value {{",
+                    snippet(cx, expr_span, "..")
+                )
+                .as_str(),
+            );
+        }
+    }
+}
+
+/// If the expression is an ExprKind::Match, check if the scrutinee has a significant drop that may
+/// have a surprising lifetime.
+fn has_significant_drop_in_scrutinee<'tcx, 'a>(cx: &'a LateContext<'tcx>, expr: &'tcx Expr<'tcx>) -> Option<Span> {
+    let mut helper = SigDropHelper::new(cx);
+    match expr.kind {
+        ExprKind::Match(match_expr, _, _source) => {
```

Comment:
```suggestion
        ExprKind::Match(match_expr, _, _) => {
```

I don't think there is a need to give the unused `span` a name.

Reply:
Done!